apps/rush-lib/src/cli/utilities/npm/NpmLinkManager.ts

```diff
@@ -0,0 +1,323 @@
+      readPackageTree(this._rushConfiguration.commonTempFolder,
+        (error: Error, npmPackage: readPackageTree.PackageNode) => {
+        if (error) {
+          reject(error);
+        } else {
+          try {
+            const commonRootPackage: NpmPackage = NpmPackage.createFromNpm(npmPackage);
+
+            const commonPackageLookup: PackageLookup = new PackageLookup();
+            commonPackageLookup.loadTree(commonRootPackage);
+
+            const rushLinkJson: IRushLinkJson = { localLinks: {} };
+
+            for (const rushProject of this._rushConfiguration.projects) {
+              console.log(os.EOL + 'LINKING: ' + rushProject.packageName);
+              this._linkProject(rushProject, commonRootPackage, commonPackageLookup, rushLinkJson);
+            }
+
+            console.log(`Writing "${this._rushConfiguration.rushLinkJsonFilename}"`);
+            JsonFile.save(rushLinkJson, this._rushConfiguration.rushLinkJsonFilename);
+
+            resolve();
+          } catch (error) {
+            reject(error);
+          }
+        }
+      });
+    });
+  }
+
+  /**
+   * This is called once for each local project from Rush.json.
+   * @param project             The local project that we will create symlinks for
+   * @param commonRootPackage   The common/temp/package.json package
+   * @param commonPackageLookup A dictionary for finding packages under common/temp/node_modules
+   * @param rushConfiguration   The rush.json file contents
+   * @param rushLinkJson        The common/temp/rush-link.json output file
+   * @param options             Command line options for "rush link"
+   */
+  private _linkProject(
+    project: RushConfigurationProject,
+    commonRootPackage: NpmPackage,
+    commonPackageLookup: PackageLookup,
+    rushLinkJson: IRushLinkJson): void {
+
+    let commonProjectPackage: NpmPackage | undefined =
+      commonRootPackage.getChildByName(project.tempProjectName) as NpmPackage;
+    if (!commonProjectPackage) {
+      // Normally we would expect the temp project to have been installed into the common\node_modules
+      // folder.  However, if it was recently added, "rush install" doesn't technically require
+      // this, as long as its dependencies can be found at the root of the NPM shrinkwrap file.
+      // This avoids the need to run "rush generate" unnecessarily.
+
+      // Example: "project1"
+      const unscopedTempProjectName: string = Utilities.parseScopedPackageName(project.tempProjectName).name;
+
+      // Example: "C:\MyRepo\common\temp\projects\project1
+      const extractedFolder: string = path.join(this._rushConfiguration.commonTempFolder,
+        RushConstants.rushTempProjectsFolderName, unscopedTempProjectName);
+
+      // Example: "C:\MyRepo\common\temp\projects\project1.tgz"
+      const tarballFile: string = path.join(this._rushConfiguration.commonTempFolder,
+        RushConstants.rushTempProjectsFolderName, unscopedTempProjectName + '.tgz');
```

Comment:
This logic should be changed

Reply:
We probably don't need to read the tarball anymore as the package.json is expected to already exist at the end of rush install

---
In reply to: [152137246](https://github.com/Microsoft/web-build-tools/pull/446#discussion_r152137246) [](ancestors = 152137246)

Reply:
Will change this later as well

---
In reply to: [152137590](https://github.com/Microsoft/web-build-tools/pull/446#discussion_r152137590) [](ancestors = 152137590,152137246)